### PR TITLE
Allow timeouts to be set for the ldap dial and handle the handshake w…

### PR DESCRIPTION
**Pull request by Nitecon**
This pull request supports Dial timeouts and also supports handling tls handshake via tls client which returns an error to the user if there is in fact an error with the tls handshake.

### CODE REVIEW

**Comment by Nitecon** [comment]
to use this and configure a timeout to be used the following can be used for documentation purposes also:

    timeout := 60 * time.Second
    flag.Set("ldapConnectionTimeout", string(timeout))

    /* For TLS Connections: Example */
    tlsConfig := tls.Config{InsecureSkipVerify:true}
    conn, err = ldap.DialTLS("tcp", server, &tlsConfig)

    /* For standard connections: Example */
    conn, err = ldap.Dial("tcp", server)

The default timeout is set to 60s as part of the default value on the flag.

File: conn.go [hunk 1]
```diff
@@ -68,9 +76,14 @@ func Dial(network, addr string) (*Conn, error) {
 // DialTLS connects to the given address on the given network using tls.Dial
 // and then returns a new Conn for the connection.
 func DialTLS(network, addr string, config *tls.Config) (*Conn, error) {
-	c, err := tls.Dial(network, addr, config)
-	if err != nil {
-		return nil, NewError(ErrorNetwork, err)
+	dc, err := net.DialTimeout(network, addr, ldapTimeout)
+	if err != nil{
+		return nil, err
```
Comment (johnweldon) [inline conn.go]: Can we continue using `NewError` here for consistency?  Maybe just like it was before:

```
return nil, NewError(ErrorNetwork, err)
```

File: conn.go [hunk 2]
```diff
@@ -68,9 +76,14 @@ func Dial(network, addr string) (*Conn, error) {
...
+	}
+	c := tls.Client(dc, config)
+	err = c.Handshake()
+	if err != nil{
+		return nil, err
```
Comment (johnweldon) [inline conn.go]: And the same thing right here?

**Comment by johnweldon** [comment]
Thanks @Nitecon, this looks great.  Can you update the error return?  Then we can merge this.

**Comment by Nitecon** [comment]
Updated.  Also updated the way the timeout string is set with flags to be more human readable and functional when using config files.

**Comment by Nitecon** [comment]
Updated examples above to reflect the change on timeout setting to now be:
    flag.Set("ldapConnectionTimeout", "10s")
or
    flag.Set("ldapConnectionTimeout", "2m")
etc

**Comment by johnweldon** [comment]
Fantastic, thanks @Nitecon

**Comment by deoxxa** [comment]
Reading a globally-defined flag is a super weird way to implement this. Not only that, but it actually causes problems for programs that use flags themselves. I've had to patch this out locally, and I'd bet good money that there are other people who will have to do the same thing.

**Comment by Nitecon** [comment]
The only reason for implementing it that way was to not cause a breaking change to the functions that may already be implemented.  I do use many flags in my app by itself which is partially the reason for the very verbose flag name.  I do not disagree that it's a much less efficient way but it was the only way to add settable timeout functionality without breaking / changing existing apps.  What kind of issues are you having that you had to patch it out?

**Comment by deoxxa** [comment]
It's actually not the only way to implement this without breaking API compatibility. For example, I've refactored this approach in go-ldap/ldap#18 to use a package-level variable instead of piggybacking on the functionality of the `flag` package.

As for the practical problems that this approach has, allow me to defer to the documentation for the `flag` package. [The documentation for `Parse()`](https://golang.org/pkg/flag/#Parse) states that it _"Must be called after all flags are defined and before flags are accessed by the program"_. Since you're running `Parse()` in `init()`, this guarantee can't be held by any programs using this library. [The example for the `flag` package](https://golang.org/pkg/flag/#example_) _explicitly_ warns against running `Parse()` in `init()`.

Here's an annoying feature of this approach:

```
➜  ldapreader git:(master) ./ldapreader
# starting up
ldapreader: error: required flag --ldap_url not provided, try --help
➜  ldapreader git:(master) ./ldapreader --help
Usage of ./ldapreader:
  -ldapConnectionTimeout="60s": Default connection timeout for ldap
```

By taking over flag parsing, this change breaks the help output of my program.

Here's another annoying thing:

```
➜  ldapreader git:(master) ./ldapreader --ldap_url x
flag provided but not defined: -ldap_url
Usage of ./ldapreader:
  -ldapConnectionTimeout="60s": Default connection timeout for ldap
```

Again, by taking over flag parsing, this change breaks not only the help output, but also the main functionality of my program.

**Comment by deoxxa** [comment]
I just realised that I'm coming off really snarky here, and I'd like to make it clear that I'm not laying blame on anybody or saying anyone did anything wrong. I just disagree with the technical implementation of this change and would like to see a less invasive method used :)

**Comment by johnweldon** [comment]
@deoxxa thanks for the explanation and clarification of your intent :)

It seems reasonable to avoid an approach that explicitly goes against the documentation for `flag`, thanks for catching that.  I didn't connect the dots when I did the code review.

@Nitecon; does the package variable approach seem workable for your needs?  It basically means moving the definition and parsing of flags out to the applications using this library, and making the consumers of this library responsible for setting the package level variable, instead of in the library itself, which does seem cleaner to me too.

**Pull request merged**